File: DS_ASSIGNMENT_1/Assignment_1/dynamicarray.cpp
#include <bits/stdc++.h>
using namespace std;

class DynamicArray {
private:
    vector<int> arr;

public:
    DynamicArray() {}

    void insert(int index, int value) {
        if (index < 0 || index > arr.size()) {
            cout << "Invalid index" << endl;
            return;
        }
        arr.insert(arr.begin() + index, value);
    }

    void deleteAtIndex(int index) {
        if (index < 0 || index >= arr.size()) {
            cout << "Invalid index" << endl;
            return;
        }
        arr.erase(arr.begin() + index);
    }

    int getSize() {
        return arr.size();
    }

    // Return true if the dynamic array is empty, false otherwise
    bool isEmpty() {
        return arr.empty();
    }

    // Rotate the dynamic array by k positions to the right
    void rotateRight(int k) {
        if (arr.empty() || k <= 0) {
            return;
        }
        k = k % arr.size();
        if (k == 0) {
            return;
        }
        rotate(arr.begin(), arr.end() - k, arr.end());
    }

    // Reverse the dynamic array
    void reverse() {
        std::reverse(arr.begin(), arr.end());
    }

    // Append an element to the end of the dynamic array
    void append(int value) {
        arr.push_back(value);
    }

    // Prepend an element to the beginning of the dynamic array
    void prepend(int value) {
        arr.insert(arr.begin(), value);
    }

    // Merge two dynamic arrays into a single dynamic array
    DynamicArray* merge(DynamicArray* other) {
        DynamicArray* merged = new DynamicArray();
        merged->arr.reserve(arr.size() + other->arr.size());
        merged->arr.insert(merged->arr.end(), arr.begin(), arr.end());
        merged->arr.insert(merged->arr.end(), other->arr.begin(), other->arr.end());
        return merged;
    }

    // Interleave two dynamic arrays into a single dynamic array
    DynamicArray* interleave(DynamicArray* other) {
        DynamicArray* interleaved = new DynamicArray();
        int i = 0, j = 0;
        int size1 = arr.size(), size2 = other->arr.size();

        while (i < size1 && j < size2) {
            interleaved->arr.push_back(arr[i++]);
            interleaved->arr.push_back(other->arr[j++]);
        }

        while (i < size1) {
            interleaved->arr.push_back(arr[i++]);
        }

        while (j < size2) {
            interleaved->arr.push_back(other->arr[j++]);
        }

        return interleaved;
    }

    // Return the middle element of the dynamic array
    int getMiddleElement() {
        if (arr.empty()) {
            cout << "Dynamic array is empty" << endl;
            return -1;
        }
        return arr[arr.size() / 2];
    }

    // Return the index of the first occurrence of the specified element
    int indexOf(int value) {
        for (int i = 0; i < arr.size(); i++) {
            if (arr[i] == value) {
                return i;
            }
        }
        return -1;
    }

    // Split the dynamic array into two dynamic arrays at the specified index
    DynamicArray* splitAtIndex(int index) {
        if (index < 0 || index > arr.size()) {
            cout << "Invalid index" << endl;
            return nullptr;
        }

        DynamicArray* secondArray = new DynamicArray();
        secondArray->arr.assign(arr.begin() + index, arr.end());
        arr.erase(arr.begin() + index, arr.end());

        return secondArray;
    }

    // Resizing the array with a custom factor
    void resize(int factor) {
        if (factor <= 0) {
            cout << "Invalid factor" << endl;
            return;
        }
        arr.resize(arr.size() * factor);
    }

    // Print the dynamic array
    void print() {
        for (int num : arr) {
            cout << num << " ";
        }
        cout << endl;
    }
};

int main() {
    DynamicArray arr;
    arr.append(1);
    arr.append(2);
    arr.append(3);
    arr.append(4);
    arr.print(); 

    arr.insert(2, 5);
    arr.print(); 

    arr.deleteAtIndex(2);
    arr.print();

    cout << "Size: " << arr.getSize() << endl; 
    cout << "Is empty? " << (arr.isEmpty() ? "Yes" : "No") << endl; 

    arr.rotateRight(2);
    arr.print(); 

    arr.reverse();
    arr.print(); 

    arr.prepend(0);
    arr.print(); 

    DynamicArray arr2;
    arr2.append(5);
    arr2.append(6);
    arr2.append(7);

    DynamicArray* merged = arr.merge(&arr2);
    merged->print(); 

    DynamicArray* interleaved = arr.interleave(&arr2);
    interleaved->print(); 

    cout << "Middle element: " << arr.getMiddleElement() << endl; 

    cout << "Index of 3: " << arr.indexOf(3) << endl;

    DynamicArray* secondArray = arr.splitAtIndex(2);
    arr.print(); 
    secondArray->print(); 

    arr.resize(3);
    arr.print(); 

    delete merged;
    delete interleaved;
    delete secondArray;

    return 0;
}

File: DS_ASSIGNMENT_1/Assignment_1/dynamicarray.py
class DynamicArray:
    def __init__(self):
        self.arr = []

    def insert(self, index, value):
        if index < 0 or index > len(self.arr):
            print("Invalid index")
            return
        self.arr.insert(index, value)

    def delete_at_index(self, index):
        if index < 0 or index >= len(self.arr):
            print("Invalid index")
            return
        del self.arr[index]

    def get_size(self):
        return len(self.arr)

    def isempty(self):
        return len(self.arr) == 0

    def rotate_right(self, k):
        if not self.arr or k <= 0:
            return
        k %= len(self.arr)
        self.arr = self.arr[-k:] + self.arr[:-k]

    def reverse(self):
        self.arr.reverse()

    def append(self, value):
        self.arr.append(value)

    def prepend(self, value):
        self.arr.insert(0, value)

    def merge(self, other):
        merged = DynamicArray()
        merged.arr = self.arr + other.arr
        return merged

    def interleave(self, other):
        interleaved = DynamicArray()
        i, j = 0, 0
        while i < len(self.arr) and j < len(other.arr):
            interleaved.append(self.arr[i])
            interleaved.append(other.arr[j])
            i += 1
            j += 1
        while i < len(self.arr):
            interleaved.append(self.arr[i])
            i += 1
        while j < len(other.arr):
            interleaved.append(other.arr[j])
            j += 1
        return interleaved

    def get_middle_element(self):
        if not self.arr:
            print("Dynamic array is empty")
            return -1
        return self.arr[len(self.arr) // 2]

    def index_of(self, value):
        if value in self.arr:
            return self.arr.index(value)
        return -1

    def split_at_index(self, index):
        if index < 0 or index > len(self.arr):
            print("Invalid index")
            return None
        second_array = DynamicArray()
        second_array.arr = self.arr[index:]
        self.arr = self.arr[:index]
        return second_array

    def resize(self, factor):
        if factor <= 0:
            print("Invalid factor")
            return
        self.arr = self.arr * factor

    def print_array(self):
        print(" ".join(map(str, self.arr)))

arr = DynamicArray()
arr.append(1)
arr.append(2)
arr.append(3)
arr.append(4)
arr.print_array()

arr.insert(2, 5)
arr.print_array()

arr.delete_at_index(2)
arr.print_array()

print("Size:", arr.get_size())
print("Is empty?", "Yes" if arr.isempty() else "No")

arr.rotate_right(2)
arr.print_array()

arr.reverse()
arr.print_array()

arr.prepend(0)
arr.print_array()

arr2 = DynamicArray()
arr2.append(5)
arr2.append(6)
arr2.append(7)

merged = arr.merge(arr2)
merged.print_array()

interleaved = arr.interleave(arr2)
interleaved.print_array()

print("Middle element:", arr.get_middle_element())

print("Index of 3:", arr.index_of(3))

second_array = arr.split_at_index(2)
arr.print_array()
second_array.print_array()

arr.resize(3)
arr.print_array()
